### PR TITLE
[API][Web][SIMS #1399 ]Student view assessment and appeal

### PR DESCRIPTION
**Pull request by dheepak-aot**
**Student Acceptance Criteria**
- [x] Once an application has reached "assessment" display the assessment tables (similar to a ministry user)
image
- [x] Hide the requests table when it's empty
image
- [x] Do not display the offering change requests
- [x] In the "your assessment" table do NOT display the ministry notes
image
- [x] when viewing a request form in the your assessment table display only the information the student submitted (same as top table)
- [x] on click of "view assessment" display the same table and buttons as #906
image
- [x] do NOT display the certificate number
image
- [x] Display "How to receive your payment" panel with content and hyperlink
image

### CODE REVIEW

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 1]
```diff
@@ -273,4 +287,124 @@ export class AssessmentControllerService {
       });
     return finalAward;
   }
+
+  /**
+   * Get all requests related to an assessments for a student
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:292]: `an assessments`?

File: sources/packages/api/src/route-controllers/assessment/assessment.students.controller.ts [hunk 2]
```diff
@@ -110,9 +112,47 @@ export class AssessmentStudentsController extends BaseController {
...
   ): Promise<AwardDetailsAPIOutDTO> {
     return this.assessmentControllerService.getAssessmentAwardDetails(
       assessmentId,
+      userToken.studentId,
+      false,
+    );
+  }
+
+  /**
+   * Get all requests related to an assessments for a student
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.students.controller.ts:125]: `an assessments`

File: sources/packages/api/src/services/student-appeal/student-appeal.service.ts [hunk 3]
```diff
@@ -147,7 +149,13 @@ export class StudentAppealService extends RecordDataModelService<StudentAppeal>
...
           );
         }),
-      )
+      );
+    if (studentId) {
+      query.andWhere("application.student.id = :studentId", {
+        studentId,
+      });
+    }
+    const queryResult = await query
```
Comment (ann-aot) [inline sources/packages/api/src/services/student-appeal/student-appeal.service.ts:158]: I think you can use the same `query` object. no need of assigning to a new variable

File: sources/packages/web/src/components/aest/AppealRequestsApprovalForm.vue [hunk 4]
```diff
@@ -38,8 +38,13 @@ export default {
...
       default: false,
     },
+    showApprovalDetails: {
+      type: Boolean,
+      required: false,
+      default: true,
+    },
   },
-  setup(_props: any, context: SetupContext) {
+  setup(_props: any, context) {
```
Comment (ann-aot) [inline sources/packages/web/src/components/aest/AppealRequestsApprovalForm.vue:47]: no need of `any`

File: sources/packages/web/src/components/students/StudentAssessmentDetails.vue [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  props: {
+    applicationId: {
+      type: Number,
+      required: true,
+    },
+  },
+  setup(props) {
+    const router = useRouter();
+    // The assessment trigger types for which the request form is visible to student.
+    const studentAssessmentRequestTypes = ref<AssessmentTriggerType[]>([
```
Comment (ann-aot) [inline sources/packages/web/src/components/students/StudentAssessmentDetails.vue:35]: looks like it is only used in the history table, can't we add this logic in the history component?

File: sources/packages/web/src/components/aest/students/assessment/History.vue [hunk 6]
```diff
@@ -107,8 +107,16 @@ export default {
       type: Number,
       required: true,
     },
+    // Assessment trigger types for which request for is available to view.
+    // When the value is default then request form is viewable for all
+    // possible trigger types.
+    viewRequestTypes: {
+      type: Object as PropType<AssessmentTriggerType[]>,
```
Comment (ann-aot) [inline sources/packages/web/src/components/aest/students/assessment/History.vue:114]: I think we do have an Array type.

File: sources/packages/web/src/components/aest/students/assessment/History.vue [hunk 7]
```diff
@@ -159,6 +168,21 @@ export default {
       }
     };
 
+    // Decide to show the request form for the provided assessment trigger types.
+    // Extending the existing logic to show request form for provided trigger types.
+    const canShowViewRequest = (
```
Comment (ann-aot) [inline sources/packages/web/src/components/aest/students/assessment/History.vue:173]: can you explain me this function?

File: sources/packages/web/src/components/aest/students/assessment/Request.vue [hunk 8]
```diff
@@ -81,8 +81,13 @@ export default {
       type: Number,
       required: true,
     },
+    showWhenEmpty: {
```
Comment (ann-aot) [inline sources/packages/web/src/components/aest/students/assessment/Request.vue:84]: if the show of the base container dependents on this property, then can't the logic be performed by the parent while calling this child? pls let me know if I am missing anything?

File: sources/packages/web/src/components/students/StudentAssessmentDetails.vue [hunk 9]
```diff
@@ -0,0 +1,65 @@
+<template>
+  <request-assessment
+    :applicationId="applicationId"
+    :showWhenEmpty="false"
```
Comment (ann-aot) [inline sources/packages/web/src/components/students/StudentAssessmentDetails.vue:4]: mm....what is the use of this?`:showWhenEmpty="false"`

File: sources/packages/web/src/components/students/StudentAssessmentDetails.vue [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  />
+</template>
+<script lang="ts">
+import { defineComponent, ref } from "vue";
+import { useRouter } from "vue-router";
+import { AssessmentTriggerType } from "@/types";
+import { StudentRoutesConst } from "@/constants/routes/RouteConstants";
+import HistoryAssessment from "@/components/aest/students/assessment/History.vue";
+import RequestAssessment from "@/components/aest/students/assessment/Request.vue";
+export default defineComponent({
```
Comment (ann-aot) [inline sources/packages/web/src/components/students/StudentAssessmentDetails.vue:21]: pls add a line before export

File: sources/packages/web/src/views/student/StudentAppealRequests.vue [hunk 11]
```diff
@@ -0,0 +1,80 @@
+<template>
+  <v-container>
```
Comment (ann-aot) [inline sources/packages/web/src/views/student/StudentAppealRequests.vue:2]: mm.. can't we use student container?

File: sources/packages/web/src/views/student/StudentAssessmentAward.vue [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    </v-card>
+    <v-card class="p-4 my-3">
+      <v-row>
+        <v-col>
+          <h2 class="category-header-medium primary-color">
+            How to receive your award
+          </h2>
+          <p class="default-color my-3">
+            Please visit the
+            <span class="font-weight-bold"
```
Comment (ann-aot) [inline sources/packages/web/src/views/student/StudentAssessmentAward.vue:41]: pls, avoid bootstrap bold. we have custom classes already in place

**Review by ann-aot — COMMENTED** [review]
Good Work @dheepak-aot  added some comments

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:292]: The comment was copied from the controller method. But I fixed the description in all places.

File: sources/packages/api/src/services/student-appeal/student-appeal.service.ts (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/api/src/services/student-appeal/student-appeal.service.ts:158]: I need the query result to assign to mapFromRawAndEntities.

image

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 13]
```diff
@@ -155,7 +169,7 @@ export class AssessmentControllerService {
    */
   private populateDisbursementAwardValues(
     disbursementSchedules: DisbursementSchedule[],
-    includeDocumentNumber = false,
+    includeDocumentNumber = true,
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:172]: Is it not safe to keep it false as it was? If the parameter is not provided the value will not be included by mistake.

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 14]
```diff
@@ -273,4 +287,124 @@ export class AssessmentControllerService {
...
   }
+
+  /**
+   * Get all requests related to an assessments for a student
+   * application, i.e, this will fetch all pending and denied
+   * student appeals for an application or possible application
+   * exceptions that will prevent the assessment to proceed till
+   * they are approved, for instance, when a document is uploaded
+   * and need to be reviewed.
+   * @param applicationId application number.
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:298]: A minor one but can we please change "application number" for "application id"?

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 15]
```diff
@@ -273,4 +287,124 @@ export class AssessmentControllerService {
...
+   * @param studentId applicant student.
+   * @param  includeOnlyAppeals then only student appeals are returned and
+   * other assessment types must not be available when this property is true.
+   * @returns assessment requests or exceptions for a student application.
+   */
+
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:304]: Please remove the extra blank line.

File: sources/packages/web/src/components/students/StudentAssessmentDetails.vue (hunk 5 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/students/StudentAssessmentDetails.vue:35]: This object must be residing at the consumer end of the component to choose what are the trigger types which are available to show request form. 

Every time parent doesn't have to provide it as the history component will show all possible request form on default.
Only if parent chooses to show only selected trigger types then they will have to supply the values.

File: sources/packages/web/src/components/aest/students/assessment/History.vue [hunk 16]
```diff
@@ -159,6 +168,21 @@ export default {
...
+      data: AssessmentHistorySummaryAPIOutDTO,
+    ): boolean => {
+      const defaultShowViewRequest = showViewRequestDefault(data);
+      if (props.viewRequestTypes.length) {
+        return (
+          defaultShowViewRequest &&
+          props.viewRequestTypes.includes(data.triggerType)
+        );
+      }
+      return showViewRequestDefault(data);
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/web/src/components/aest/students/assessment/History.vue:183]: I would start prefer the method name to start with action. I can change to showDefaultViewRequest.

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 17]
```diff
@@ -273,4 +287,124 @@ export class AssessmentControllerService {
...
+        studentAppealId: assessment.studentAppeal?.id,
+        applicationExceptionId: assessment.application.applicationException?.id,
+        studentScholasticStandingId: assessment.studentScholasticStanding?.id,
+      }),
+    );
+    // Add unsuccessful scholastic standing to the top of the list, if present.
+    // For unsuccessful scholastic standing, status is always "completed" and
+    // "createdAt" is "submittedDate".
+    if (unsuccessfulScholasticStanding) {
+      history.unshift({
```
Comment (henryoforeh-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:399]: For my understanding, was unshift used here to ensure the most recent record is displayed first ?

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 18]
```diff
@@ -273,4 +287,124 @@ export class AssessmentControllerService {
...
+   */
+
+  async getRequestedAssessmentSummary(
+    applicationId: number,
+    studentId?: number,
+    includeOnlyAppeals = false,
+  ): Promise<RequestAssessmentSummaryAPIOutDTO[]> {
+    const requestAssessmentSummary: RequestAssessmentSummaryAPIOutDTO[] = [];
+    // Requested offering changes and application exceptions must not be included when includeOnlyAppeals is true.
+    if (!includeOnlyAppeals) {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:312]: It seems the queries inside this "if" are executed without checking if the student has access to the application provided as a parameter, is it right? If it is, we should either check the authorization before the queries are executed or include the authorization in the query (for students only).

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts [hunk 19]
```diff
@@ -273,4 +287,124 @@ export class AssessmentControllerService {
...
+        id: appeals.id,
+        submittedDate: appeals.submittedDate,
+        status: appeals.status,
+        requestType: RequestAssessmentTypeAPIOutDTO.StudentAppeal,
+      }));
+    return requestAssessmentSummary.concat(studentAppealArray);
+  }
+
+  /**
+   * Method to get history of assessments for an application,
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:358]: I would prefer to remove from the comments the sentence part that states of the code is like: "Method", "Class", "Endpoint", and "Property". Without it is as meaningful, as shown below.
```
Get the history of assessments for an application
```

File: sources/packages/web/src/components/aest/students/assessment/Request.vue (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/aest/students/assessment/Request.vue:84]: This component is built in a way that it just takes the applicationId and it makes the API call inside the component. So parent doesn't have any control over the size of the data table. 

In student side of this story, the requests table needs to be hidden when the table is empty. Hence added the logic to component.

File: sources/packages/web/src/components/students/StudentAssessmentDetails.vue (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/students/StudentAssessmentDetails.vue:4]: As mentioned in the above comment, in student side of this story, the requests table needs to be hidden when the table is empty.

File: sources/packages/web/src/views/student/StudentAppealRequests.vue (hunk 11 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/student/StudentAppealRequests.vue:2]: Sure. Missed it.

File: sources/packages/web/src/views/student/StudentAssessmentAward.vue (hunk 12 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/student/StudentAssessmentAward.vue:41]: Okay. I will create an exclusive class to make the font weight bold.

File: sources/packages/web/src/components/aest/AppealRequestsApprovalForm.vue [hunk 20]
```diff
@@ -38,8 +38,13 @@ export default {
...
       default: false,
     },
+    showApprovalDetails: {
+      type: Boolean,
+      required: false,
+      default: true,
+    },
   },
-  setup(_props: any, context: SetupContext) {
+  setup(_props, context) {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/AppealRequestsApprovalForm.vue:47]: Not part of the PR but the context is missing the type here.

File: sources/packages/web/src/components/aest/students/assessment/History.vue [hunk 21]
```diff
@@ -107,8 +107,16 @@ export default {
...
+    // When the value is default then request form is viewable for all
+    // possible trigger types.
+    viewRequestTypes: {
+      type: Array as PropType<AssessmentTriggerType[]>,
+      required: false,
+      default: [] as AssessmentTriggerType[],
+    },
   },
-  setup(props: any, context: SetupContext) {
+  setup(props, context) {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/assessment/History.vue:119]: What is the benefit of removing the type here? `context: SetupContext`

File: sources/packages/web/src/components/aest/students/assessment/Request.vue [hunk 22]
```diff
@@ -81,8 +81,13 @@ export default {
...
+      type: Boolean,
+      required: false,
+      default: true,
+    },
   },
-  setup(props: any, context: SetupContext) {
+  setup(props, context) {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/assessment/Request.vue:90]: Please keep the `context: SetupContext`.

File: sources/packages/web/src/components/students/StudentAssessmentDetails.vue (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/students/StudentAssessmentDetails.vue:35]: I believe that it does not need to be declared reactive since its value will never change. It can also be added directly to the template (but either way is fine).

File: sources/packages/web/src/views/student/StudentAppealRequests.vue [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+      required: true,
+    },
+    appealId: {
+      type: Number,
+      required: true,
+    },
+  },
+  setup(props) {
+    const studentAppealRequests = ref<StudentAppealRequest[]>([]);
+    const appealStatus = ref<StudentAppealStatus>(StudentAppealStatus.Pending);
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/StudentAppealRequests.vue:57]: I believe that it can be declared as below. It will be typed in the same way.
```ts
const appealStatus = ref(StudentAppealStatus.Pending);
```

File: sources/packages/web/src/views/student/StudentAssessmentAward.vue [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+            >
+            to collect your awarded funds. It may take time to process if you do
+            not see it immediately. Please contact NSLSC if you have additional
+            questions about receiving your funds.
+          </p>
+          <v-btn
+            color="primary"
+            prepend-icon="fa:fas fa-up-right-from-square"
+            href="https://protege-secure.csnpe-nslsc.canada.ca/en/public/register-account"
+            target="_blank"
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/StudentAssessmentAward.vue:52]: Please include the attribute `rel="noopener"`. It is the Sonarcloud recommendation for security issues while redirecting to an external website.

**Review by andrewsignori-aot — COMMENTED** [review]
Gret work, the assessment is looking pretty good 👍 
Please take a look at the comments.

File: sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts (hunk 17 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/assessment/assessment.controller.service.ts:395]: It was used to add elements in the beginning of the array and this logic was existing already.

**Review by henryoforeh-aot — APPROVED** [review]

File: sources/packages/web/src/views/student/StudentAppealRequests.vue [hunk 25]
```diff
@@ -0,0 +1,78 @@
+<template>
+  <full-page-container>
```
Comment (ann-aot) [inline sources/packages/web/src/views/student/StudentAppealRequests.vue:2]: we have `<student-page-container`. u have used it in `sources/packages/web/src/views/student/StudentAssessmentAward.vue`

Comment (dheepak-aot) [reply sources/packages/web/src/views/student/StudentAppealRequests.vue]: Nice catch. fixed and pushed.

**Review by ann-aot — APPROVED** [review]
Good work @dheepak-aot 👍  Thanks for doing the changes

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @dheepak-aot

**Review by andrewsignori-aot — APPROVED** [review]
Great work @dheepak-aot, thanks for doing the changes, looks good 👍
